diff --git a/parser.go b/parser.go
index f9f5ce4..fb06f29 100644
--- a/parser.go
+++ b/parser.go
@@ -2,10 +2,11 @@ package fastjson
 
 import (
 	"fmt"
-	"github.com/valyala/fastjson/fastfloat"
 	"strconv"
 	"strings"
 	"unicode/utf16"
+
+	"github.com/valyala/fastjson/fastfloat"
 )
 
 // Parser parses JSON.
@@ -29,6 +30,7 @@ type Parser struct {
 // Use Scanner if a stream of JSON values must be parsed.
 func (p *Parser) Parse(s string) (*Value, error) {
 	s = skipWS(s)
+	// todo: 这里为什么要复制，不复制行不行
 	p.b = append(p.b[:0], s...)
 	p.c.reset()
 
@@ -65,6 +67,7 @@ func (c *cache) reset() {
 }
 
 func (c *cache) getValue() *Value {
+	// 看不懂 ???
 	if cap(c.vs) > len(c.vs) {
 		c.vs = c.vs[:len(c.vs)+1]
 	} else {
@@ -73,6 +76,7 @@ func (c *cache) getValue() *Value {
 	return &c.vs[len(c.vs)-1]
 }
 
+// 跳过空白字符
 func skipWS(s string) string {
 	if len(s) == 0 || s[0] > 0x20 {
 		// Fast path.
@@ -82,10 +86,12 @@ func skipWS(s string) string {
 }
 
 func skipWSSlow(s string) string {
+	// todo: 使用查表法
 	if len(s) == 0 || s[0] != 0x20 && s[0] != 0x0A && s[0] != 0x09 && s[0] != 0x0D {
 		return s
 	}
 	for i := 1; i < len(s); i++ {
+		// todo: 使用查表法
 		if s[i] != 0x20 && s[i] != 0x0A && s[i] != 0x09 && s[i] != 0x0D {
 			return s[i:]
 		}
@@ -109,7 +115,8 @@ func (c *cache) parseValue(s string, depth int) (*Value, string, error) {
 	if depth > MaxDepth {
 		return nil, s, fmt.Errorf("too big depth for the nested JSON; it exceeds %d", MaxDepth)
 	}
-
+	// todo: 考虑传入指针
+	// todo: 使用 switch ，有可能会利用到 jump table
 	if s[0] == '{' {
 		v, tail, err := c.parseObject(s[1:], depth)
 		if err != nil {
@@ -220,6 +227,7 @@ func (c *cache) parseObject(s string, depth int) (*Value, string, error) {
 	}
 
 	if s[0] == '}' {
+		// 读到一个空对象
 		v := c.getValue()
 		v.t = TypeObject
 		v.o.reset()
@@ -282,6 +290,7 @@ func escapeString(dst []byte, s string) []byte {
 	return strconv.AppendQuote(dst, s)
 }
 
+// todo: 值得使用 avx2 加速
 func hasSpecialChars(s string) bool {
 	if strings.IndexByte(s, '"') >= 0 || strings.IndexByte(s, '\\') >= 0 {
 		return true
